src/Psalm/Internal/Type/SimpleAssertionReconciler.php

```diff
@@ -1482,7 +1483,7 @@ private static function reconcileInArray(
         Codebase $codebase,
         Union $existing_var_type,
         string $assertion
-    ) : Union {
+    ) : ?Union {
```

Comment:
In what context does Type::intersectUnionTypes is expected to return null and what happens when we return null here?

Reply:
These are the cases where haystack contains type which isn't present in the needle, like:
```
/**
 * @param int|bool $x
 * @param non-empty-list<int|string> $y
 * @return int
 */
function assertInArray($x, $y) {
    if (in_array($x, $y, true)) {
        return $x;
    }

    throw new Exception();
}
```
I missed the fact that some sort of Union instance will have to be returned eventually. I added some tests, which indeed revealed the problem. Fixed by returning TEmpty if intersection is null. Return type was reverted to `Union` (not nullable)

Reply:
Great!